sktime/forecasting/online_experts.py

```diff
@@ -0,0 +1,353 @@
+                          for prediction in expert_predictions[:, i]
+                          ])
+
+            average_loss = np.dot(self.weights, loss_vector)
+
+            instant_regret = (average_loss - loss_vector)
+            self.R += instant_regret
+            self._update_weights(low_c=low_c)
+
+    def _update_weights(self, low_c=0.01):
+        """Updates the weights on each of the experts by performing a potential
+        function update with a root-finding search. low_c represents the lower
+        bound on the window that the root finding is occuring over.
+
+        Parameters
+        ----------
+        low_c : float
+        """
+
+        # Calculating Normalizing Constant
+        R_plus = np.array(list(map(lambda x: 0 if 0 > x else x, self.R)))
+        normalizing_R = np.max(R_plus)
+
+        R_plus /= normalizing_R
+
+        low_c = low_c
+        high_c = (max(R_plus)**2)/2
+
+        def pot(c):
+            """Internal Potential Function
+
+            Parameters
+            ----------
+            low_c : float
+
+            Returns
+            -------
+            potential: float
+            """
+            return np.mean(np.exp((R_plus**2)/(2*c)))-np.e
+
+        c_t = bisection(low_c, high_c, pot)
+
+        def prob(r, c_t):
+            """Internal Probability Function
+
+            Parameters
+            ----------
+            r : float
+            c_t : float
+
+            Returns
+            -------
+            prob : float
+            """
+            return (r/c_t)*np.exp((r**2)/(2*c_t))
+
+        self.weights = np.array([prob(r, c_t) for r in R_plus])
+        self.weights /= np.sum(self.weights)
+
+
+"""
+Helper Functions
+"""
+
+
+def bisection(low, high, function, threshold=1e-8):
+    """ Uses the bisection method in order to a root of a function
+        within a specific interval
+
+    Parameters
+    ----------
+    low : float, lower bound for the interval
+    high : float, higher bound for the interval
+    function: function
+    threshold: float, abs distance from zero when the algorithm stops at
+
+    Returns
+    -------
+    mid : float, value at which the funtion takes on a value within
+          [-theshold/2,theshold]
+    """
+
+    left = low
+    right = high
+
+    if function(low) > 0:
+        left = high
+        right = low
+
+    while abs(left - right) > 1e-8:
+        mid = (left + right)/2
+        if function(mid) > 0:
+            right = mid
+        else:
+            left = mid
+
+    return (left + right)/2
+
+
+def se(actual, expected):
+    """
+    Returns the squared error between the two arguments:
+
+    Parameters
+    ----------
+    actual : float, actual value to be compared with
+    expected : float, expected value to be compared with
+
+    Returns
+    -------
+    se: squared error between the two values
+    """
+    return np.power(np.subtract(actual, expected), 2)
+
+
+def mse(actual, expected):
```

Comment:
scikit-learn has a mean_squared_error function I think, I'd find it more readable if you would write: 
np.mean((actual - expected) ** 2)

where do you use the use function?

Reply:
I see, I'll probably import that function if I were to use it. I don't use it anywhere and I'll remove it.